# RNaseViz-RNase-Sequence-Visualization-Tool

RNaseViz is a web-based application built with R Shiny, providing an interactive platform for visualizing RNase sequences, their conservation across different species, and the mutations associated with Mendelian diseases.
https://norreanea.shinyapps.io/RNaseViz/
## Features

- Selection of RNase enzymes from a predefined list.
- Visualization of sequence conservation using multiple sequence alignments.
- Interactive phylogenetic trees illustrating evolutionary relationships.
- Highlighting specific mutations and their positions within the sequences.
- Support for various RNase enzymes with associated Mendelian disease mutations.

## Usage

To use the RNaseViz application:

1. **Visualize**: From the dropdown menu labeled "Choose RNase:", select the RNase enzyme you wish to analyze and click the 'Visualize' button.
2. **Highlight Mutation**: Use the "Select organism:" dropdown to choose the organism of interest, enter a position number in the reference sequence, and click 'Highlight position' to view specific mutations.
3. **RNase info**: Provides in-depth information about the selected RNase enzyme, including the potential molecular consequences of its mutations.
4. **Residue Highlighting**: Highlights the specified residue within the sequence alignment, along with the two adjacent residues on either side, to emphasize key positions.
5. **Explore Tree**: The displayed tree structure allows you to explore how different RNase sequences relate to different species.
6. **Inspect Residues**: Select and highlight any residue within the sequence alignment for detailed examination and analysis.
7. **Download Global Alignment**: Download sequence alignments in FASTA or PDF formats for offline analysis.
![Figure2](https://github.com/user-attachments/assets/c43109ba-74c9-4e5b-98e7-6e06c615b9f2)

## Contact

For questions or support, please open an issue on this [GitHub repository](https://github.com/Norreanea/RNaseViz-RNase-Sequence-Visualization-Tool/issues).

